## Group SQL Practice

_Author: Matt Brems & Katie Sylvia_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [107]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [96]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT COUNT(*) AS num_rows
FROM stores;
"""

pd.read_sql_query(sql, engine)

num_rows
0      1973

**Prompt 1 Answer**: There are 1973 rows in the stores table in the Iowa liquor database.

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [97]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT name
FROM stores
ORDER BY name
ASC
LIMIT 4;
"""

pd.read_sql_query(sql, engine)[3:4]

name
3  3061 Cub Foods / Sioux City

**Prompt 2 Answer**: The fourth store is 3061 Cub Foods/Sioux City.

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [98]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT btl_price, bottle_qty, vendor
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor
LIMIT 3
"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty                     vendor
0     $8.99          12  Adamba Imports Int'l Inc.
1    $28.50           3      A Hardy / U.S.A. Ltd.
2    $28.50           1      A Hardy / U.S.A. Ltd.

**Prompt 3 Answer**: The top tranasaction was from Adamba Imports Int'l Inc for 12 bottles at $8.99 a bottle. The second transaction was from A Hardy / U.S.A. Ltd. for 6 bottles at $28.50 a bottle. The three transaction was from A Hardy / U.S.A. for 3 bottles at $28.50 a bottle.


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [99]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT SUM(total) AS total_amt_sold_in_feb
FROM sales
WHERE date >= '2015-02-01' 
AND date < '2015-03-01';

"""
pd.read_sql_query(sql, engine)

total_amt_sold_in_feb
0            21295350.55

**Prompt 4 Answer**: The total amount of liquor sold in February 2015 was $21,295,350.55.

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [100]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT vendor, COUNT(*) AS transaction_count,
SUM(bottle_qty) AS total_bottle_sold, SUM(total) As total_revenue

FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY transaction_count
DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

vendor  transaction_count  total_bottle_sold  \
0  Pernod Ricard USA/Austin Nichols              42338             403450   

   total_revenue  
0     7880365.88

**Prompt 5 Answer**: Pernod Richard USA/Austin Nicols had 42,338 transactions, with 403,450 bottles sold for a total revenue of $7,880,365.88.

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [101]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT btl_price - state_btl_cost as markup
FROM sales

"""

pd.read_sql_query(sql, engine)

markup
0         $5.75
1         $5.75
2         $5.75
3         $8.58
4         $2.50
...         ...
3049908   $5.75
3049909  $10.00
3049910   $5.75
3049911   $5.75
3049912  $10.00

[3049913 rows x 1 columns]

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [102]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT st.name AS store_name, 
AVG((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) As average_percentage_markup

FROM stores AS st, sales AS sa
WHERE st.store = sa.store
GROUP BY st.name
ORDER BY average_percentage_markup
DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

store_name  average_percentage_markup
0  Wal-Mart 0841 / Tipton                   0.525471

**Prompt 7 Answer**: Wal-Mart 0841 / Tipton has the largest average percentage markup with 52.55%

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [103]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT st.store, st.name AS store_name, st.store_address,  
AVG((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) As average_percentage_markup,

MIN((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) As minimum_percentage_markup,
MAX((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) As maximum_percentage_markup

FROM stores AS st, sales AS sa
WHERE st.store = sa.store
GROUP BY st.name, st.store, st.store_address
ORDER BY average_percentage_markup
DESC
LIMIT 5

"""

pd.read_sql_query(sql, engine)

store                  store_name  \
0   4013      Wal-Mart 0841 / Tipton   
1   4024  Wal-Mart 1546 / Iowa Falls   
2   4266  Wal-Mart 1683 / Shenandoah   
3   3833    Wal-Mart 3394 / Atlantic   
4   3660   Wal-Mart 2935 / Knoxville   

                                       store_address  \
0  1126 Highway 38 North\nTipton, IA 527720000\n(...   
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...   
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...   
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....   
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...   

   average_percentage_markup  minimum_percentage_markup  \
0                   0.525471                   0.498571   
1                   0.521499                   0.498462   
2                   0.518903                   0.498462   
3                   0.515094                   0.498462   
4                   0.514591                   0.498462   

   maximum_percentage_markup  
0                   6.708763  
1                   6.708763  
2                   6.708763  
3                   6.708763  
4                   6.708763

**Prompt 8 Answer**: The names of the five stores are: Wal-Mart 4013, Wal-Mart 4024, Wal-Mart 4266, Wal-Mart 3833, Wal-Mart 3660.

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [104]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT c.county, 
(SUM(s.bottle_qty)) / c.population AS bottle_per_capita,
ROUND(AVG(p.bottle_size), 1) as avg_bottle_size

FROM counties c, sales s, products p

WHERE c.county = s.county AND p.item_no = s.item
GROUP BY c.county
ORDER BY bottle_per_capita
DESC
LIMIT 9
"""

pd.read_sql_query(sql, engine)

county  bottle_per_capita  avg_bottle_size
0      Dickinson                 22            974.1
1           Polk                 15            892.6
2     Black Hawk                 15            862.9
3    Cerro Gordo                 14            986.4
4          Scott                 13            871.6
5        Johnson                 13            895.4
6           Linn                 12            886.3
7  Pottawattamie                 11            893.7
8        Kossuth                 11           1041.0

**Prompt 9 Answer**: Dickinson had the largest bottle per capita. Kossuth county sells the largest average bottle size at 1041.0ml.

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [108]:
# Use a SQL query to find the answer to the above prompt.

sql = """

select count(vendor)
from sales
where total::numeric <> (btl_price * bottle_qty)::numeric
limit 10
"""

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**: There are 0 transactions that have the value of total that is not equal to btl_price tome bottle_qty.